Tutorial 3 Convert

F20170604 opened the pull request:
Classifying Images of Clothing
Original File Link : https://github.com/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l03c01_classifying_images_of_clothing.ipynb
Swift Colab File Link : https://colab.research.google.com/drive/1XfT75glapWdCp0Zj_4mkiMrjT6odD-Gc

F20170604 committed:
Tutorial 3 Convert

Classifying Images of Clothing
Original File Link : https://github.com/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l03c01_classifying_images_of_clothing.ipynb

F20170604 committed:
Downloading process modified

F20170604 committed:
Update S4TF_Tutorial_3.ipynb

Ayush517 commented:
@dynamicwebpaige @BradLarson @rxwei Can you please let me know the modifications I should make before this is merged.
I think I might change the download process. I learnt about a better one in fast.ai notebooks.

BradLarson commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1772 @@
+        }
+      },
+      "source": [
+        "let fileBaseCommand = \"https://raw.githubusercontent.com/tensorflow/swift-models/stable/MNIST/\"\n",
```
I know you said that you'd probably rework the file loading, but this line defines a variable that doesn't seem to be used anywhere, and can probably be removed.

BradLarson commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1772 @@
+      },
+      "source": [
+        "let filemgr = FileManager.default\n",
+        "let filelist = try filemgr.contentsOfDirectory(atPath: \".\")\n",
```
I'd probably simplify this by removing the `let filemgr` line and just doing `FileManager.default.contentsOfDirectory...` here. Also the indentation of the trailing curly bracket just below here should probably be moved to the left to be aligned properly.

BradLarson commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1772 @@
+        }
+      },
+      "source": [
+        "let (trainImages, trainNumericLabels, numTrainExamples) = readMNIST(imagesFile: \"train-images-idx3-ubyte?raw=true\",\n",
```
To avoid a line-wrap here, maybe you could move `imagesFile:` parameter to a new line just below `readMNIST(`. You could make it line up with the following `labelsFile:` parameter. Same with the readMNIST below this.

BradLarson commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1772 @@
+        "let firstTrainLabel: Int = Int((trainNumericLabels[0].scalars)[0])\n",
+        "let numpyArray = firstTrainImage.makeNumpyArray().reshape(28, 28)\n",
+        "\n",
+        "//print(\"firstTrainImage type:\", type(of: firstTrainImage))\n",
```
The comments here (and in examples that follow) are probably leftover debugging comments, so those could all be removed before final commit of this notebook to remove that debugging remnant.

F20170604 replied:
I left them intentionally, so that people who want to understand better can comment them off and see what are the shapes and types being used for processing.

Ayush517 replied:
@BradLarson Should I remove them or let them be?

BradLarson replied:
Honestly, I'd remove these to focus on the core concepts being shown. If you wanted to call out the ability to inspect type, shape, etc., it might be worthwhile to have that in a separate cell with accompanying text. In this notebook, I think you're fine without the commented code.

Ayush517 commented:
There are 2 functions which I will need to directly copy from [fast.ai notebook](https://github.com/fastai/fastai_docs/blob/master/dev_swift/00_load_data.ipynb). Should I mention that I used fast.ai's notebook or let it be?

F20170604 committed:
Changes done as proposed by Brad

Updates :

1.) 2 Packages installed at the beginning of notebook
2.) 3 more imports made
3.) extension named "String" added [from fast.ai notebook]
4.) function named "downloadFile" added [from fast.ai notebook]
5.) downloading method changed. Now using original source instead of my GitHub Folder
6.) `let filemgr` line removed
7.) line-wrapping avoided
8.) debugging comments removed
9.) Minor variable name changed in an explanation

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        {
+          "output_type": "display_data",
+          "data": {
+            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsoAAAKzCAYAAAAHuytJAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzsnXnYHEW1/7+HyA6BLGQhOxBkCxJ8\nCesVZE0QBBQQkO3KBe9z4QqiYERFLz8XQBHkgmhYDChhkTVwo2ENezAJS8hGEkKAQBZCNiDs1O+P\nmal8uzLVb709W830+TzP+7xnpqu7a7q/c7p66vQ5YoyBoiiKoiiKoihJ1ml0B ...
```
1. Please use descriptive argument labels. What does `i` stand for?
2. Camel case should be used for all function base names, argument labels, variables names, and such. And the image argument should be a plural noun. Thus, `Image: ` should be changed to `images: `.
3. `predictionsArray` -> `predictions`.
4. `plotImageStats` -> `plotImageStatistics`.

rxwei replied:
Remove redundant parentheses.
```suggestion
        "    let predLabel = predictionsArray[i].makeNumpyArray()\n",
```

rxwei replied:
Use reshape on the Swift side before converting to Numpy.
```suggestion
        "    let numpyImage = images[I].reshaped(to: [28, 28]).makeNumpyArray()\n",
```

rxwei replied:
Remove redundant type annotations.
```suggestion
        "    let predictedLabelIndex = Int((predictionsArray[i].argmax().scalars)[0])\n",
```

rxwei replied:
```suggestion
        "    let correctLabelIndex = Int((correctLabels[i].scalars)[0])\n",
```

rxwei replied:
Remove parentheses around `if`/`guard`/`for`/`while` conditions. Same elsewhere.
```suggestion
        "    if correctLabelIndex != predictedLabelIndex\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        {
+          "output_type": "display_data",
+          "data": {
+            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsoAAAKzCAYAAAAHuytJAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzsnXnYHEW1/7+HyA6BLGQhOxBkCxJ8\nCesVZE0QBBQQkO3KBe9z4QqiYERFLz8XQBHkgmhYDChhkTVwo2ENezAJS8hGEkKAQBZCNiDs1O+P\nmal8uzLVb709W830+TzP+7xnpqu7a7q/c7p66vQ5YoyBoiiKoiiKoihJ1ml0B ...
```
Move `{` to the end of the previous line. Same elsewhere.

F20170604 replied:
I can't find this line. Which code segment are we talking about here?

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        }
+      },
+      "source": [
+        "public func downloadFile(_ url: String, dest: String? = nil, force: Bool = false) {\n",
```
- Avoid abbreviations in argument labels.
- `url:` should be an argument label.
```suggestion
        "public func downloadFile(url: String, destination: String? = nil, force: Bool = false) {\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+      },
+      "source": [
+        "public func downloadFile(_ url: String, dest: String? = nil, force: Bool = false) {\n",
+        "    let dest_name = dest ?? (Path.cwd/url.split(separator: \"/\").last!).string\n",
```
Use camel case for all names.
```suggestion
        "    let destName = dest ?? (Path.cwd/url.split(separator: \"/\").last!).string\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        "    print(\"Downloading \\(url)...\")\n",
+        "\n",
+        "    if let cts = Just.get(url).content {\n",
+        "        do    {try cts.write(to: URL(fileURLWithPath:dest_name))}\n",
```
```suggestion
        "        do { try cts.write(to: URL(fileURLWithPath: destName)) }\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        "\n",
+        "    if let cts = Just.get(url).content {\n",
+        "        do    {try cts.write(to: URL(fileURLWithPath:dest_name))}\n",
+        "        catch {print(\"Can't write to \\(url_dest).\\n\\(error)\")}\n",
```
```suggestion
        "        catch { print(\"Can't write to \\(url_dest).\\n\\(error)\") }\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        "\n",
+        "    print(\"Downloading \\(url)...\")\n",
+        "\n",
+        "    if let cts = Just.get(url).content {\n",
```
It is unclear what `cts` means. Use `content` instead.

```suggestion
        "    if let content = Just.get(url).content {\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+        "let filelist = try FileManager.default.contentsOfDirectory(atPath: \".\")\n",
+        "for filename in filelist {\n",
+        "        print(filename)\n",
+        "    }"
```
```suggestion
        "}"
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+      "source": [
+        "let filelist = try FileManager.default.contentsOfDirectory(atPath: \".\")\n",
+        "for filename in filelist {\n",
+        "        print(filename)\n",
```
Indent by 4.
```suggestion
        "    print(filename)\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1769 @@
+      "source": [
+        "/// Reads a file into an array of bytes.\n",
+        "func readFile(_ path: String) -> [UInt8] {\n",
+        "    let possibleFolders  = [\".\", \"MNIST\"]\n",
```
```suggestion
        "    let possibleFolders = [\".\", \"MNIST\"]\n",
```

F20170604 committed:
Changes proposed by Richard

F20170604 committed:
Redundant type annotations removed

Ayush517 commented:
@dynamicwebpaige @BradLarson @rxwei Any more changes?

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1768 @@
+        "plt.figure(figsize: [12, 12])\n",
+        "var dis = 0\n",
+        "for image in 0...24 {\n",
+        "    let TrainImage = trainImages[image]\n",
```
Please use camel case for all variable names.
`TrainImage` -> `trainImage`.

F20170604 replied:
I'm sorry. I don't know how I missed it

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1768 @@
+      "source": [
+        "plt.figure(figsize: [12, 12])\n",
+        "var dis = 0\n",
+        "for image in 0...24 {\n",
```
This is an index, not an image. Therefore something like `i` or `index` is better.

```suggestion
        "for i in 0...24 {\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1768 @@
+        "var dis = 0\n",
+        "for image in 0...24 {\n",
+        "    let TrainImage = trainImages[image]\n",
+        "    let TrainLabel = Int((trainNumericLabels[image].scalars)[0])\n",
```
`TrainLabel` -> `trainLabel`

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1768 @@
+        "var dis = 0\n",
+        "for image in 0...24 {\n",
+        "    let TrainImage = trainImages[image]\n",
+        "    let TrainLabel = Int((trainNumericLabels[image].scalars)[0])\n",
```
Every use of `.scalar` copies the entire tensor from the accelerator to the host (CPU) side. So it's better to avoid this pattern. Instead, do indexing on the tensor side and use `.scalarized()` to turn it to a scalar.
```suggestion
        "    let trainLabel = Int((trainNumericLabels[i][0].scalarized())\n",
```

F20170604 replied:
This gives me an error:

`Fatal error: Index out of range using input dim 0; input has only 0 dims: file /swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2123`

F20170604 replied:
The `trainNumericLabels[image][0]` itself has just 1 value. So whenever I .scalar it, it only copies 1 value, the same it will copy in case of `scalarized()`
I've done `Int((trainNumericLabels[image].scalars)[0])` because Int needs a binaryInteger that's why `[0]` index is used at the end so that it can tell the Int typecasting outside the bracket that only 1 value is being passed on to be converted to Int.

And wherever scalars is used in this notebook, it's the same case. Just for a single value of label

rxwei replied:
Ah right. In that case, `trainNumericLabels[i].scalarized()` should just work.

Ayush517 replied:
Should I consider this as resolved?

F20170604 replied:
So, anything else left?

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1768 @@
+        {
+          "output_type": "display_data",
+          "data": {
+            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsoAAAKzCAYAAAAHuytJAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzsnXnYHEW1/7+HyA6BLGQhOxBkCxJ8\nCesVZE0QBBQQkO3KBe9z4QqiYERFLz8XQBHkgmhYDChhkTVwo2ENezAJS8hGEkKAQBZCNiDs1O+P\nmal8uzLVb709W830+TzP+7xnpqu7a7q/c7p66vQ5YoyBoiiKoiiKoihJ1ml0B ...
```
Always wrap backquotes around code in text or comments.
```suggestion
        "`classifier(x)` returns a list of lists of probabilities, one for each image in the batch of data. Grab the predictions for our (first) image in the batch:"
```

rxwei replied:
Indent by 4 at each line break.
```suggestion
        "                Tensor<Int32>(correctPredictions).sum().scalarized())\n",
```

rxwei commented on S4TF_Tutorial_3.ipynb:
```diff
@@ -0,0 +1,1768 @@
+        "public extension String {\n",
+        "    @discardableResult\n",
+        "    func shell(_ args: String...) -> String\n",
+        "    {\n",
```
Put `{` at the end of the previous line.

F20170604 committed:
All changes made except one

I haven't updated the scalarized() part in this update. It's giving me an error.
